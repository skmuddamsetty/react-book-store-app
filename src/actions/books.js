// import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_BOOK
export const addBook = book => ({
  type: 'ADD_BOOK',
  book
});
// REMOVE_BOOK
export const removeBook = ({ id } = {}) => ({
  type: 'REMOVE_BOOK',
  id
});
// EDIT_BOOK

export const editBook = (id, updates) => ({
  type: 'EDIT_BOOK',
  id,
  updates
});

export const startAddBook = (bookData = {}) => {
  return dispatch => {
    const { description = '', note = '', price = 0, createdAt = 0 } = bookData;
    const book = { description, note, price, createdAt };
    return database
      .ref('books')
      .push(book)
      .then(ref => {
        dispatch(addBook({ id: ref.key, ...book }));
      })
      .catch();
  };
};
