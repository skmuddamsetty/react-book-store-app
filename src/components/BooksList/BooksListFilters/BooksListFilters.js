import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByDate,
  sortByPrice
} from '../../../actions/filters';

export class BooksListFilters extends Component {
  onTextChange = event => {
    this.props.setTextFilter(event.target.value);
  };

  onSelectChange = event => {
    if (event.target.value === 'price') {
      this.props.sortByDate();
    } else if (event.target.value === 'date') {
      this.props.sortByPrice();
    }
  };

  render() {
    return (
      <div>
        <input
          type='text'
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select
          valueArray={this.props.filters.sortBy}
          onChange={this.onSelectChange}
        >
          <option value='date'>Date</option>
          <option value='price'>Price</option>
        </select>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return { filters: state.filters };
};

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByPrice: () => dispatch(sortByPrice())
});

export default connect(mapStateToProps, mapDispatchToProps)(BooksListFilters);
