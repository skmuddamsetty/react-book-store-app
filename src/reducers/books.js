// Books reducer
const booksReducerDefaultState = [];

export const booksReducer = (state = booksReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_BOOK':
      //   return state.concat(action.book);
      return [...state, action.book];
    case 'REMOVE_BOOK':
      return state.filter(book => book.id !== action.id);
    case 'EDIT_BOOK':
      return state.map(book => {
        if (book.id === action.id) {
          return { ...book, ...action.updates };
        } else {
          return book;
        }
      });
    default:
      return state;
  }
};
