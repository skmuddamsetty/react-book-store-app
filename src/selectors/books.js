// get Visible Books
export const getVisibleBooks = (
  books,
  { text, sortBy, startDate, endDate }
) => {
  return books
    .filter(book => {
      const startDateMatch =
        typeof startDate !== 'number' || book.createdAt >= startDate;
      const endDateMatch =
        typeof endDate !== 'number' || books.createdAt <= endDate;
      const textMatch = book.description
        .toLowerCase()
        .includes(text.toLowerCase());
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'price') {
        return a.price < b.price ? -1 : 1;
      } else {
        return 0;
      }
    });
};
