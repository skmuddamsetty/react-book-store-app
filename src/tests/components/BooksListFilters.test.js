import React from 'react';
import { shallow } from 'enzyme';
import { BooksListFilters } from '../../components/BooksList/BooksListFilters/BooksListFilters';
import { books } from '../fixtures/books';
import toJSON from 'enzyme-to-json';
import { filters, altFilters } from '../fixtures/filters';

let setTextFilter, sortByDate, sortByPrice, wrapper;

beforeEach(() => {
  setTextFilter = jest.fn();
  sortByDate = jest.fn();
  sortByPrice = jest.fn();
  wrapper = shallow(
    <BooksListFilters
      filters={filters}
      setTextFilter={setTextFilter}
      sortByDate={sortByDate}
      sortByPrice={sortByPrice}
    />
  );
});
test('should render BooksListFilters properly', () => {
  expect(wrapper).toMatchSnapshot();
});
