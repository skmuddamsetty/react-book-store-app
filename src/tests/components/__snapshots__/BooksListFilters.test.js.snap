// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BooksListFilters properly 1`] = `
ShallowWrapper {
  "length": 1,
}
`;
